python/cudf/cudf/core/column/interval.py

```diff
@@ -0,0 +1,91 @@
+# Copyright (c) 2018-2021, NVIDIA CORPORATION.
+import pyarrow as pa
+import cudf
+from cudf.core.column import StructColumn
+
+
+class IntervalColumn(StructColumn):
+    def __init__(
+        self,
+        dtype,
+        mask=None,
+        size=None,
+        offset=0,
+        null_count=None,
+        children=(),
+        closed="right",
+    ):
+
+        super().__init__(
+            data=None,
+            dtype=dtype,
+            mask=mask,
+            size=size,
+            offset=offset,
+            null_count=null_count,
+            children=children,
+        )
+        self._closed = closed
+
+    @property
+    def closed(self):
+        if self._closed in ["left", "right", "neither", "both"]:
+            return self._closed
+        else:
+            raise ValueError("closed value is not valid")
+
+    @classmethod
+    def from_arrow(self, data):
+        new_col = super().from_arrow(data.storage)
+        size = len(data)
+        dtype = cudf.core.dtypes.IntervalDtype.from_arrow(data.type)
+        mask = data.buffers()[0]
+        if mask is not None:
+            mask = cudf.utils.utils.pa_mask_buffer_to_mask(mask, len(data))
+
+        offset = data.offset
+        null_count = data.null_count
+        children = new_col.children
+        closed = dtype.closed
+
+        return IntervalColumn(
+            size=size,
+            dtype=dtype,
+            mask=mask,
+            offset=offset,
+            null_count=null_count,
+            children=children,
+            closed=closed,
+        )
+
+    def to_arrow(self):
+        typ = self.dtype.to_arrow()
+        return pa.ExtensionArray.from_storage(typ, super().to_arrow())
+
+    def from_struct_column(self, closed="right"):
+        return IntervalColumn(
+            size=self.size,
+            dtype=cudf.core.dtypes.IntervalDtype(
+                self.dtype.fields["left"], closed
+            ),
+            mask=self.mask,
+            offset=self.offset,
+            null_count=self.null_count,
+            children=self.children,
```

Comment:
Since we're passing in the offset here, I believe we should be using `self.base_mask` and `self.base_children`

Reply:
All good, I'll change this 👍🏾 Is this something that we always do when we pass in offset?

Reply:
Yes, `self.mask` and `self.children` are copy constructed when there's an offset in order to make it logically 0. I.e. take a column like: `['abc', null, 'de']`. If you sliced it with `slice(1, None)`, then `offset` would be `1` and the column would effectively look like: `[null, 'de']`.

In order to accomplish this, the way we handle things is that `base_children` would still look like:
```
child[0]: ['a', 'b', 'c', 'd', 'e'] # characters
child[1]: [0, 3, 3, 5] # offsets
```
but we'd want `children` to look like:
```
child[0]: ['d', 'e'] # characters
child[1]: [0, 0, 2] # offsets
```

So we actually need to calculate new `offsets` in this situation, which isn't free both from a compute and memory perspective.

Libcudf handles these situations internally so we want to always pass the relevant `offset` / `base_*` into libcudf, whereas if we're copying to something like PyArrow on the host it often makes the amount of data being moved much smaller so its worth it to calculate the new reduced columns.